strawberry/permission.py

```diff
@@ -1,21 +1,115 @@
-    from strawberry.types.info import Info
+    from strawberry.field import StrawberryField
+    from strawberry.types import Info
 
 
-class BasePermission:
+class BasePermission(abc.ABC):
     """
     Base class for creating permissions
     """
 
     message: Optional[str] = None
 
+    error_extensions: Optional[dict[str, str]] = None
+
+    error_class: Type[GraphQLError] = StrawberryGraphQLError
+
+    _schema_directive: Optional[StrawberrySchemaDirective] = None
+
+    @abc.abstractmethod
     def has_permission(
         self, source: Any, info: Info, **kwargs
     ) -> Union[bool, Awaitable[bool]]:
         raise NotImplementedError(
             "Permission classes should override has_permission method"
         )
+
+    def raise_error(self) -> None:
+        """
+        Default error raising for permissions. This can be overridden to customize the behavior.
+        """
+
+        # Instantiate error class
+        error = self.error_class(self.message)
+
+        if self.error_extensions:
+            # If no custom error class is used, use the standard GraphQLError so we can add the extensions
+            if not error:
+                error = GraphQLError(self.message)
+            # Add our extensions to the error
+            error.extensions.update(self.error_extensions)
+
+        # If none of the new attributes were used, fall back to prevent breaking changes
+        if not error:
+            raise PermissionError(self.message)
+
+        raise error
+
+    @property
+    def schema_directive(self) -> Optional[StrawberrySchemaDirective]:
+        if not self._schema_directive:
+            self._schema_directive = StrawberrySchemaDirective(
+                self.__class__.__name__,
+                self.__class__.__name__,
+                [Location.FIELD_DEFINITION],
+                [],
+            )
+        return self._schema_directive
+
+
+class PermissionExtension(FieldExtension):
+    def __init__(self, permissions: List[BasePermission]):
+        self.permissions = permissions
+
+    def apply(self, field: StrawberryField) -> None:  # nocov
+        """Applies all of the permission directives to the schema"""
+        for permission in self.permissions:
+            if permission.schema_directive:
+                cast(List, field.directives).append(permission.schema_directive)
+
+    def resolve(
+        self, next: Callable[..., Any], source: Any, info: Info, **kwargs
+    ) -> Any:
+        """
+        Checks if the permission should be accepted and
+        raises an exception if not
+        """
+        for permission in self.permissions:
+            if not permission.has_permission(source, info, **kwargs):
+                permission.raise_error()
+
+    async def resolve_async(
+        self, next: Callable[..., Any], source: Any, info: Info, **kwargs
+    ) -> Any:
+        for permission in self.permissions:
+            has_permission: bool
+
+            has_permission = await await_maybe(
```

Comment:
Maybe instead of using `await_maybe` we can wrap non async permissions into async functions during `__init__` to cache those? The main reason is that the `inspect.isawaitable` call is very expensive.

Reply:
But I just noticed that we use that on schema_converter for the resolver function. Probably it will be better to try to optimize all of those cases in a separated PR

Reply:
Actually, nevermind this comment. I thought about [this issue](https://github.com/graphql-python/graphql-core/issues/54), but it seems to not be a major issue for newer python versions anymore.

I made a quick test of returning 1 million objects with an async resolver that returned an int. It took ~52sec with the current `await_maybe` and ~51sec when I removed it. It also took ~51sec when I replaced `inspect.isawaitable` by [graphql-core's own implementation](https://github.com/graphql-python/graphql-core/blob/main/src/graphql/pyutils/is_awaitable.py). I'm using python 3.11

Even using an `asyncio.Future`, which [is the worst case to check](https://github.com/python/cpython/blob/main/Lib/inspect.py#L454) had the same differences.

So apparently there isn't too much overhead for newer python versions anymore.

Reply:
Do you think it makes sense to `asyncio.gather` all async permissions here? It shouldn't break anything since permissions should not have side-effects, but make these checks run faster.

Reply:
We talked about performance drawbacks of using gather in the #performance channel today. I think it's best to leave it as is for now.